frame/contracts/proc-macro/src/lib.rs

```diff
@@ -579,18 +635,33 @@ fn expand_functions(
 ///
 /// The implementation on `()` can be used in places where no `Ext` exists, yet. This is useful
 /// when only checking whether a code can be instantiated without actually executing any code.
+///
+/// # Generating Documentation
+///
+/// Passing `doc` attribute to the macro (like `#[define_env(doc)]`) will make it also expand
+/// additional `pallet_contracts::wasm::runtime::seal0`, `pallet_contracts::wasm::runtime::seal1`,
+/// `...` modules each having its `Doc` trait containing methods holding documentation for every
+/// defined host function.
+///
+/// To build up these docs, run:
+///
+/// ```nocompile
+/// 	cargo doc --no-deps --document-private-items
+/// ```
 #[proc_macro_attribute]
 pub fn define_env(attr: TokenStream, item: TokenStream) -> TokenStream {
-	if !attr.is_empty() {
-		let msg = "Invalid `define_env` attribute macro: expected no attributes: `#[define_env]`.";
+	if !attr.is_empty() && !(attr.to_string() == "doc".to_string()) {
```

Comment:
I think we should rename `doc` to `emit_doc`. I think this makes it clearer what is happening. I won't die on that hill if you disagree, though.

Reply:
to me, concise variant is more elegant and could hardly be misinterpreted

Reply:
But isn't it the same attribute name used for the doc comments itself?

Reply:
But isn't it the same attribute name used for rust doc comments itself? To be that seems confusing. Also, why doesn't this PR add this attribute?

Reply:
> But isn't it the same attribute name used for rust doc comments itself? 

Yeah, that's exactly the point. The idea is to make it short, appropriate and familiar to the user: at the end of the day, it is for generating the rustdoc. I don't see it is confusing, as here `doc` attr is used as input to `define_env(doc)` rather than solo (as when it's for a single doc comment) 

> Also, why doesn't this PR add this attribute to the define_env invocation in runtime.rs?

I was thinking to leave it off by default. But could be done vice versa, why not. Added.

Reply:
> I don't see it is confusing, as here doc attr is used as input to define_env(doc) rather than solo (as when it's for a single doc comment)

Fair enough.

> I was thinking to leave it off by default. But could be done vice versa, why not. Added.

Most important point for me is that the standard docs.rs builds of this crate contain this documentation. Therefore it should be on.

Reply:
> Most important point for me is that the standard docs.rs builds of this crate contain this documentation. Therefore it should be on.

Gotcha.  The only objection comes to my mind is that it probably should be turned off for production use not to bloat runtime size in vain. Should we add a note to the macro docs at least?

Reply:
It should not bloat the runtime size as it is dead code and hence removed by the compiler. At least when using lto.

But you have a point there. I think you should just add `#[cfg(doc)]]` to the emitted module. Then this code will be compiled out for every target but rustdoc.